### PR TITLE
CPLP-1657 add update service provider details endpoint

### PR DESCRIPTION
**Pull request by Phil91**
With this PR [CPLP-1657](https://jira.catena-x.net/browse/CPLP-1657) is resolved.

### CODE REVIEW

**Comment by Phil91** [comment]
I don't understand why TestCoverage is shown as 76.7% the shown as untested code is covered.

**Comment by evegufy** [comment]
> I don't understand why TestCoverage is shown as 76.7% the shown as untested code is covered.

You mean the lines at the bottom here are covered?: https://sonarcloud.io/code?id=catenax-ng_product-portal-backend&pullRequest=274&selected=catenax-ng_product-portal-backend%3Asrc%2Fportalbackend%2FPortalBackend.DBAccess%2FRepositories%2FCompanyRepository.cs

**Comment by Phil91** [comment]
> > I don't understand why TestCoverage is shown as 76.7% the shown as untested code is covered.
> 
> You mean the lines at the bottom here are covered?: https://sonarcloud.io/code?id=catenax-ng_product-portal-backend&pullRequest=274&selected=catenax-ng_product-portal-backend%3Asrc%2Fportalbackend%2FPortalBackend.DBAccess%2FRepositories%2FCompanyRepository.cs

yes. they are covered with multiple test cases: https://github.com/catenax-ng/product-portal-backend/pull/274/files#diff-44c04d8737c33165098313143f0fd9098f1cbae600c82523df86f22af958188c

**Comment by evegufy** [comment]
> > > I don't understand why TestCoverage is shown as 76.7% the shown as untested code is covered.
> > 
> > 
> > You mean the lines at the bottom here are covered?: https://sonarcloud.io/code?id=catenax-ng_product-portal-backend&pullRequest=274&selected=catenax-ng_product-portal-backend%3Asrc%2Fportalbackend%2FPortalBackend.DBAccess%2FRepositories%2FCompanyRepository.cs
> 
> yes. they are covered with multiple test cases: https://github.com/catenax-ng/product-portal-backend/pull/274/files#diff-44c04d8737c33165098313143f0fd9098f1cbae600c82523df86f22af958188c

Sometimes I see the following in the logs of sonarcloud https://github.com/catenax-ng/product-portal-backend/actions/runs/3264348820/jobs/5364854600#step:7:2500 , could it be connected to that?

**Comment by Phil91** [comment]
> > > > I don't understand why TestCoverage is shown as 76.7% the shown as untested code is covered.
> > > 
> > > 
> > > You mean the lines at the bottom here are covered?: https://sonarcloud.io/code?id=catenax-ng_product-portal-backend&pullRequest=274&selected=catenax-ng_product-portal-backend%3Asrc%2Fportalbackend%2FPortalBackend.DBAccess%2FRepositories%2FCompanyRepository.cs
> > 
> > 
> > yes. they are covered with multiple test cases: https://github.com/catenax-ng/product-portal-backend/pull/274/files#diff-44c04d8737c33165098313143f0fd9098f1cbae600c82523df86f22af958188c
> 
> Sometimes I see the following in the logs of sonarcloud https://github.com/catenax-ng/product-portal-backend/actions/runs/3264348820/jobs/5364854600#step:7:2500 , could it be connected to that?

okay, that's interesting never recognized that. Looks like the build server can't connect to docker properly. Since we need this to start our Integration and Database Tests it's very likely that the percentage is connected with that.

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs [hunk 1]
```diff
@@ -175,6 +176,109 @@ public async Task GetServiceProviderCompanyDetailsAsync_WithInvalidServiceProvid
...
+        //Arrange
+        var serviceProviderDetailData = new ServiceProviderDetailData("https://www.service-url.com");
+        var serviceProviderDetailDataId = Guid.NewGuid();
+        var sut = _fixture.Create<ServiceProviderBusinessLogic>();
+            
+        //Act
+        async Task Action() => await sut.UpdateServiceProviderCompanyDetailsAsync(serviceProviderDetailDataId, serviceProviderDetailData, Guid.NewGuid().ToString()).ConfigureAwait(false);
+
+        //Assert
+        A.CallTo(() => _portalRepositories.SaveAsync()).MustNotHaveHappened();
```
Comment (ntruchsess) [inline tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs:207]: that assert cannot fail as Action() has not been executed yet.

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs [hunk 2]
```diff
@@ -175,6 +176,109 @@ public async Task GetServiceProviderCompanyDetailsAsync_WithInvalidServiceProvid
...
```
Comment (ntruchsess) [inline tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs:224]: that assert cannot fail as Action() has not been executed yet.

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs [hunk 3]
```diff
@@ -175,6 +176,109 @@ public async Task GetServiceProviderCompanyDetailsAsync_WithInvalidServiceProvid
...
+    {
+        //Arrange
+        var serviceProviderDetailData = new ServiceProviderDetailData("http://www.service-url.com");
+        var sut = _fixture.Create<ServiceProviderBusinessLogic>();
+            
+        //Act
+        async Task Action() => await sut.UpdateServiceProviderCompanyDetailsAsync(ExistingServiceProviderCompanyDetailId, serviceProviderDetailData, IamUserId).ConfigureAwait(false);
+
+        //Assert
+        A.CallTo(() => _portalRepositories.SaveAsync()).MustNotHaveHappened();
```
Comment (ntruchsess) [inline tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs:240]: that assert cannot fail as Action() has not been executed yet.

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs [hunk 4]
```diff
@@ -175,6 +176,109 @@ public async Task GetServiceProviderCompanyDetailsAsync_WithInvalidServiceProvid
...
+    {
+        //Arrange
+        var serviceProviderDetailData = new ServiceProviderDetailData(string.Empty);
+        var sut = _fixture.Create<ServiceProviderBusinessLogic>();
+            
+        //Act
+        async Task Action() => await sut.UpdateServiceProviderCompanyDetailsAsync(ExistingServiceProviderCompanyDetailId, serviceProviderDetailData, IamUserId).ConfigureAwait(false);
+
+        //Assert
+        A.CallTo(() => _portalRepositories.SaveAsync()).MustNotHaveHappened();
```
Comment (ntruchsess) [inline tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs:257]: that assert cannot fail as Action() has not been executed yet.

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs [hunk 5]
```diff
@@ -175,6 +176,109 @@ public async Task GetServiceProviderCompanyDetailsAsync_WithInvalidServiceProvid
...
+    {
+        //Arrange
+        var serviceProviderDetailData = new ServiceProviderDetailData("https://www.super-duper-long-url-which-is-actually-to-long-to-be-valid-but-it-is-not-long-enough-yet-so-add-a-few-words.com");
+        var sut = _fixture.Create<ServiceProviderBusinessLogic>();
+            
+        //Act
+        async Task Action() => await sut.UpdateServiceProviderCompanyDetailsAsync(ExistingServiceProviderCompanyDetailId, serviceProviderDetailData, IamUserId).ConfigureAwait(false);
+
+        //Assert
+        A.CallTo(() => _portalRepositories.SaveAsync()).MustNotHaveHappened();
```
Comment (ntruchsess) [inline tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs:274]: that assert cannot fail as Action() has not been executed yet.

**Review by ntruchsess — CHANGES_REQUESTED** [review]
I did a small refactoring that allows to differentiate in between 'NotFound' and 'Forbidden'. 
Please pull before fixing the comments.

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs (hunk 1 above)
Comment (ntruchsess) [inline tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs:207]: that assert cannot fail as Action() has been declared but not yet been executed. It actually being scheduled to run by awaiting Assert.ThrowsAsync(Action).

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs (hunk 1 above)
Comment (Phil91) [reply tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs]: order fixed

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs (hunk 3 above)
Comment (Phil91) [reply tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs]: order fixed

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs (hunk 4 above)
Comment (Phil91) [reply tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs]: order fixed

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs (hunk 5 above)
Comment (Phil91) [reply tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs]: order fixed

File: tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs (hunk 1 above)
Comment (Phil91) [reply tests/administration/Administration.Service.Tests/BusinessLogic/ServiceProviderBusinessLogicTest.cs]: order fixed

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**